functions/gateway/services/nats_service.go

```diff
@@ -228,47 +226,100 @@ func (s *NatsService) ConsumeMsg(ctx context.Context, workers int) error {
-
-							// Delete the existing future events
-							_, err = BulkDeleteEventsFromWeaviate(context.Background(), weaviateClient, eventIdsToDelete)
-							if err != nil {
-								log.Printf("Failed to delete existing future events with EventSourceId %s: %v", eventSourceId, err)
-								// Continue with processing even if deletion fails
+						// Match criteria: Name, Location, and Time (within tolerance)
+						nameMatch := existingEvent.Name == newEvent.EventTitle
+						locationMatch := existingEvent.Address == newEvent.EventLocation
+
+						// For time matching, parse the new event's time IN THE EVENT'S TIMEZONE
+						// The scraped time string is in local time (same timezone as the existing event)
+						timeMatch := false
+						var timeDiff int64
+						if newEvent.EventStartTime != "" {
+							// Use the existing event's timezone to parse the new event time
+							// This ensures we're comparing apples to apples (both in the same timezone context)
+							newEventTime, err := time.ParseInLocation("2006-01-02T15:04:05", newEvent.EventStartTime, &existingEvent.Timezone)
+							if err == nil {
+								// Compare Unix timestamps (both in UTC now) - exact match required
+								timeDiff = abs(existingEvent.StartTime - newEventTime.Unix())
+								timeMatch = timeDiff == 0 // Exact time match
 							} else {
-								log.Printf("Successfully deleted %d existing future events with EventSourceId %s", len(eventIdsToDelete), eventSourceId)
+								log.Printf("Time parse error: %v", err)
 							}
 						}
+
+						if nameMatch && locationMatch && timeMatch {
+							// This is a match - preserve the existing event and skip inserting the new one
+							preservedEventIds[existingEvent.Id] = true
+							newEventIndicesToSkip[j] = true
+							break
+						} else {
+							continue
+						}
 					}
 				}
 
-				// TODO: bring this back with properly extracted events
+				// Step 5 & 6: For each PreservedEvent, DO NOT delete
+				// Only delete events that are NOT in the preserved list
+				eventIdsToDelete := make([]string, 0)
+				for _, existingEvent := range existingEvents {
+					if !preservedEventIds[existingEvent.Id] {
+						eventIdsToDelete = append(eventIdsToDelete, existingEvent.Id)
+					}
+				}
 
-				if len(events) == 0 {
-					log.Printf("No events extracted from %s", seshuJob.NormalizedUrlKey)
+				if len(eventIdsToDelete) > 0 {
+					log.Printf("Deleting %d events that no longer exist at source", len(eventIdsToDelete))
+					_, err = BulkDeleteEventsFromWeaviate(context.Background(), weaviateClient, eventIdsToDelete)
+					if err != nil {
+						log.Printf("Failed to delete obsolete events: %v", err)
+					} else {
+						log.Printf("Successfully deleted %d obsolete events", len(eventIdsToDelete))
+					}
 				} else {
-					log.Printf("Extracted %d events from %s", len(events), seshuJob.NormalizedUrlKey)
+					log.Printf("No obsolete events to delete")
```

Comment:
Please check that this variable is in scope (I didn't test it)

```suggestion
					log.Printf("No events to delete for : %s", seshuJob.NormalizedUrlKey)
```